lib-src/libcs/getint.3
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1994 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" See <cmu_copyright.h> for use and distribution information.
.\" 
.\" 
.\" HISTORY
.\" $Log: getint.3,v $
.\" Revision 4.1  1997/01/08 21:53:53  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:28  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:32  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:23  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:54:12  moore]
.\"
.\" $EndLog$
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" $Log: getint.3,v $
.\" Revision 4.1  1997/01/08 21:53:53  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:28  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:32  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:23  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:54:12  moore]
.\"
.\" Revision 1.1.1.2  1995/10/18  14:54:12  moore
.\" 	Created libcs man(3) pages from mach entries
.\"
.\" Revision 1.2  90/12/12  15:51:17  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\" 	Revised for 4.3.
.\" 
.\" 23-Oct-82  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Error or EOF on standard input now causes default value to be used.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH GETINT 3 10/23/82
.CM 3
.SH "NAME"
getint, getlong, getshort \- ask user to type an integer value
.SH "SYNOPSIS"
.B
#include <stdio.h>
.br
.B
#include <ctype.h>
.sp
.nf
.B
int getint (prompt,min,max,defalt);
.B
char *prompt;
.B
int min,max,defalt;
.sp
.B
long getlong (prompt,min,max,defalt);
.B
char *prompt;
.B
long min,max,defalt;
.sp
.B
short getshort (prompt,min,max,defalt);
.B
char *prompt;
.B
short min,max,defalt;
.fi
.SH "DESCRIPTION"
.I
Getint,
.I
getlong,
and
.I
getshort
ask the user to type in an integer.
.sp
They begin by printing the string
.I
prompt
as a message to the user.
The user then types in a number.
If the number is valid and is within the range
.I
min
to
.I
max,
then it is returned as the value of
.IR "getint (getlong, getshort)" .
If it is invalid or is out of range, then an error message
is printed and the prompt-and-response cycle is repeated.
If the user types just a carriage return, then the value
.I
defalt
is assumed.
.sp
.I
Getint, getlong,
and
.I
getshort
are identical, except for the type of the parameters and the results.
.SH "LIBRARY NOTES"
You must link against /usr/local/lib/libcs.a to define this rotine.
.SH "SEE ALSO"
.IR getbool (3), 
.IR getstr (3), 
.IR gethex (3), 
.IR getstab (3), 
.IR getdouble (3), 
etc.
.br
.IR intarg (3), 
.IR longarg (3), 
.IR shortarg (3)


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


lib-src/libcs/getint.c
#ifndef _BLURB_
#define _BLURB_
/*
 * This code was originally part of the CMU SCS library "libcs".
 * A branch of that code was taken in May 1996, to produce this
 * standalone version of the library for use in Coda and Odyssey
 * distribution.  The copyright and terms of distribution are
 * unchanged, as reproduced below.
 *
 * Copyright (c) 1990-96 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */

static char *rcsid = "$Header: getint.c,v 1.1 96/06/03 19:00:33 satya Exp $";
#endif /*_BLURB_*/

/*  getint --  prompt user for int
 *
 *  Usage:  i = getint (prompt,min,max,defalt)
 *	int i,min,max,defalt;
 *	char *prompt;
 *
 *  Getint prints the message:  prompt  (min to max)  [defalt]
 *  and accepts a line of input from the user.  If the input
 *  is not null or numeric, an error message is printed; otherwise,
 *  the value is converted to an int (or the value "defalt" is
 *  substituted if the input is null).  Then, the value is
 *  checked to ensure that is lies within the range "min" to "max".
 *  If it does not, an error message is printed.  As long as
 *  errors occur, the cycle is repeated; when a legal value is
 *  entered, this value is returned by getint.
 *  On error or EOF in the standard input, the default is returned.
 *
 */

#include <stdio.h>
#include <ctype.h>

#include "libcs.h"

int getint (prompt,min,max,defalt)
int min,max,defalt;
const char *prompt;
{
	char input [200];
	register char *p;
	register int i = 0, err;

	fflush (stdout);
	do {

		fprintf (stderr,"%s  (%d to %d)  [%d]  ",prompt,min,max,defalt);
		fflush (stderr);

		if (gets (input) == NULL) {
			i = defalt;
			err = (i < min || max < i);
		}
		else {
			err = 0;
			for (p=input; *p && (isdigit(*p) || *p == '-' || *p == '+'); p++) ;
	
			if (*p) {		/* non-numeric */
				err = 1;
			} 
			else {
				if (*input)	i = atol (input);
				else		i = defalt;
				err = (i < min || max < i);
			}
		}

		if (err) fprintf (stderr,"Must be a number between %d and %d\n",
		min,max);
	} 
	while (err);

	return (i);
}
